File: go/simulate_test.go
package afi

import (
	"math/big"
	"testing"

	"github.com/ethereum/go-ethereum/common"
)

func TestLookupBalanceSlot_BaseWETH(t *testing.T) {
	slot, ok := LookupBalanceSlot(8453, common.HexToAddress("0x4200000000000000000000000000000000000006"))
	if !ok {
		t.Fatal("expected slot for Base WETH")
	}
	if slot != 3 {
		t.Errorf("expected slot 3 for WETH, got %d", slot)
	}
}

func TestLookupBalanceSlot_Unknown(t *testing.T) {
	_, ok := LookupBalanceSlot(8453, common.HexToAddress("0x000000000000000000000000000000000000dEaD"))
	if ok {
		t.Error("expected ok=false for unknown token")
	}
}

func TestRegisterBalanceSlot(t *testing.T) {
	addr := common.HexToAddress("0x000000000000000000000000000000000000dEaD")
	defer delete(balanceSlots[8453], addr)

	RegisterBalanceSlot(8453, addr, 42)
	got, ok := LookupBalanceSlot(8453, addr)
	if !ok || got != 42 {
		t.Errorf("expected slot 42, got %d ok=%v", got, ok)
	}
}

func TestBuildBalanceOverride(t *testing.T) {
	holder := common.HexToAddress("0xAaAaAaaAaAaAaAaAaAaAAAAAaaaAaAaAaAaAaAaA")
	token := common.HexToAddress("0x4200000000000000000000000000000000000006") // Base WETH, slot 3
	amount := new(big.Int).SetUint64(1000000)

	override, err := BuildBalanceOverride(8453, token, holder, amount)
	if err != nil {
		t.Fatal(err)
	}
	acct, ok := override[token]
	if !ok {
		t.Fatal("override missing token entry")
	}
	if len(acct.StateDiff) != 1 {
		t.Errorf("expected 1 stateDiff entry, got %d", len(acct.StateDiff))
	}
	// Derived slot must be keccak256(holder || 3)
	expectedSlot := mappingSlot(holder, 3)
	val, ok := acct.StateDiff[expectedSlot]
	if !ok {
		t.Fatal("derived slot missing in stateDiff")
	}
	if val != common.BigToHash(amount) {
		t.Errorf("amount mismatch: got %s want %s", val.Hex(), common.BigToHash(amount).Hex())
	}
}

func TestBuildBalanceOverride_UnknownToken(t *testing.T) {
	holder := common.HexToAddress("0xAaAaAaaAaAaAaAaAaAaAAAAAaaaAaAaAaAaAaAaA")
	token := common.HexToAddress("0x000000000000000000000000000000000000bEEF")

	_, err := BuildBalanceOverride(8453, token, holder, big.NewInt(1))
	if err == nil {
		t.Fatal("expected error for unknown token")
	}
}

func TestMappingSlot_Deterministic(t *testing.T) {
	h := common.HexToAddress("0x1234567890123456789012345678901234567890")
	a := mappingSlot(h, 9)
	b := mappingSlot(h, 9)
	if a != b {
		t.Error("mappingSlot must be deterministic")
	}
	c := mappingSlot(h, 10)
	if a == c {
		t.Error("different slots must produce different keys")
	}
}

func TestSimulateOpts_RequiresClient(t *testing.T) {
	_, err := SimulateRoute(nil, SimulateOpts{Amount: big.NewInt(1)})
	if err == nil {
		t.Fatal("expected error for nil RPCClient")
	}
}

func TestSimulateOpts_RequiresAmount(t *testing.T) {
	// Trick: pass a stub RPC client that won't be called since amount check is first
	_, err := SimulateRoute(nil, SimulateOpts{Amount: big.NewInt(0)})
	if err == nil {
		t.Fatal("expected error for zero amount")
	}
}

File: nodejs/src/__tests__/simulate.test.ts
import { describe, expect, it } from "vitest"
import { keccak256, encodeAbiParameters, getAddress, pad, type Hex } from "viem"
import { mappingSlot, simulateRoute, detectBalanceSlot } from "../simulate.js"
import { lookupBalanceSlot, registerBalanceSlot } from "../token-slots.js"
import type { Address } from "../types.js"

// A publicClient stub whose balanceOf reflects the overridden slot value
// (faithful EVM behavior), so detectBalanceSlot matches the first probed slot.
// quote() calls return an encoded (outputAsset, amountOut) tuple.
function fakeClient(asset: Address, amountOut: bigint) {
  return {
    call: async ({ data, stateOverride }: any) => {
      if (typeof data === "string" && data.startsWith("0x70a08231")) {
        return { data: stateOverride[0].stateDiff[0].value as Hex }
      }
      return {
        data: encodeAbiParameters(
          [{ type: "address" }, { type: "uint256" }],
          [asset, amountOut],
        ),
      }
    },
  } as never
}

describe("lookupBalanceSlot", () => {
  it("returns slot 3 for Base WETH", () => {
    const slot = lookupBalanceSlot(8453, "0x4200000000000000000000000000000000000006")
    expect(slot).toBe(3)
  })

  it("is case-insensitive", () => {
    const slot = lookupBalanceSlot(8453, "0x4200000000000000000000000000000000000006".toUpperCase() as Address)
    expect(slot).toBe(3)
  })

  it("returns undefined for unknown token", () => {
    const slot = lookupBalanceSlot(8453, "0x000000000000000000000000000000000000dEaD")
    expect(slot).toBeUndefined()
  })
})

describe("registerBalanceSlot", () => {
  it("adds runtime entries readable by lookup", () => {
    const addr = "0x000000000000000000000000000000000000bEEF" as Address
    registerBalanceSlot(8453, addr, 42)
    expect(lookupBalanceSlot(8453, addr)).toBe(42)
  })
})

describe("mappingSlot", () => {
  it("matches keccak256(holder . slot) by hand", () => {
    const holder = getAddress("0x1234567890abcdef1234567890abcdef12345678")
    const slot = 9
    const expected = keccak256(
      encodeAbiParameters(
        [{ type: "address" }, { type: "uint256" }],
        [holder, BigInt(slot)],
      ),
    )
    expect(mappingSlot(holder, slot)).toBe(expected)
  })

  it("accepts a bigint slot", () => {
    const holder = getAddress("0x1234567890abcdef1234567890abcdef12345678")
    expect(mappingSlot(holder, 9n)).toBe(mappingSlot(holder, 9))
  })

  it("is deterministic", () => {
    const h = getAddress("0x1234567890123456789012345678901234567890")
    expect(mappingSlot(h, 5)).toBe(mappingSlot(h, 5))
  })

  it("varies with slot", () => {
    const h = getAddress("0x1234567890123456789012345678901234567890")
    expect(mappingSlot(h, 1)).not.toBe(mappingSlot(h, 2))
  })
})

describe("simulateRoute input validation", () => {
  it("rejects zero amount", async () => {
    await expect(
      simulateRoute({
        publicClient: {} as never,
        chainId: 8453,
        quoterAddress: "0x0000000000000000000000000000000000000001",
        asset: "0x4200000000000000000000000000000000000006",
        amount: 0n,
        stepsEncoded: "0x00",
      }),
    ).rejects.toThrow(/amount must be > 0/)
  })

  it("auto-detects the slot for an unknown token and caches it", async () => {
    const asset = getAddress("0x000000000000000000000000000000000000cafe")
    expect(lookupBalanceSlot(8453, asset)).toBeUndefined()

    const res = await simulateRoute({
      publicClient: fakeClient(asset, 999n),
      chainId: 8453,
      quoterAddress: "0x0000000000000000000000000000000000000001",
      asset,
      amount: 1n,
      stepsEncoded: "0x00",
    })

    expect(res.reverted).toBe(false)
    expect(res.amountOut).toBe(999n)
    // slot detected on the first probe (0) and cached
    expect(lookupBalanceSlot(8453, asset)).toBe(0)
  })
})

describe("detectBalanceSlot", () => {
  it("detects and registers the backing slot", async () => {
    const token = "0x000000000000000000000000000000000000F00d" as Address
    expect(lookupBalanceSlot(1, token)).toBeUndefined()

    const slot = await detectBalanceSlot(fakeClient(token, 0n), 1, token)
    expect(slot).toBe(0)
    expect(lookupBalanceSlot(1, token)).toBe(0)
  })

  it("throws when no slot matches within maxSlot", async () => {
    // client that never echoes the sentinel → no slot ever matches
    const client = {
      call: async () => ({ data: pad("0x00" as Hex, { size: 32 }) }),
    } as never
    await expect(
      detectBalanceSlot(client, 1, "0x000000000000000000000000000000000000dEaD", 3),
    ).rejects.toThrow(/not detected/)
  })

  it("skips a slot whose eth_call throws and matches a later one", async () => {
    const sentinel = pad(`0x${(0xdeadbeefdeadbeefn).toString(16)}` as Hex, { size: 32 })
    let n = 0
    const client = {
      call: async () => {
        n++
        if (n === 1) throw new Error("rpc hiccup on slot 0")
        return { data: sentinel } // slot 1 echoes the sentinel
      },
    } as never
    const token = "0x000000000000000000000000000000000000Ab1e" as Address
    expect(await detectBalanceSlot(client, 1, token)).toBe(1)
  })
})

describe("simulateRoute revert handling", () => {
  const asset = getAddress("0x00000000000000000000000000000000000bEEf1")

  function throwingQuoteClient(err: unknown) {
    registerBalanceSlot(8453, asset, 0) // pre-register so detection is skipped
    return { call: async () => { throw err } } as never
  }

  async function run(err: unknown) {
    return simulateRoute({
      publicClient: throwingQuoteClient(err),
      chainId: 8453,
      quoterAddress: "0x0000000000000000000000000000000000000001",
      asset,
      amount: 1n,
      stepsEncoded: "0x00",
    })
  }

  it("extracts revert data from err.cause.data", async () => {
    const res = await run({ cause: { data: "0xdeadbeef" } })
    expect(res.reverted).toBe(true)
    expect(res.revertData).toBe("0xdeadbeef")
  })

  it("falls back to err.data", async () => {
    const res = await run({ data: "0xfeed" })
    expect(res.reverted).toBe(true)
    expect(res.revertData).toBe("0xfeed")
  })

  it("leaves revertData undefined when the error carries none", async () => {
    const res = await run(new Error("plain revert"))
    expect(res.reverted).toBe(true)
    expect(res.revertData).toBeUndefined()
  })

  it("treats an empty quote result as reverted", async () => {
    const empty = getAddress("0x00000000000000000000000000000000000bEEf2")
    registerBalanceSlot(8453, empty, 0) // skip detection
    const client = { call: async () => ({ data: undefined }) } as never
    const res = await simulateRoute({
      publicClient: client,
      chainId: 8453,
      quoterAddress: "0x0000000000000000000000000000000000000001",
      asset: empty,
      amount: 1n,
      stepsEncoded: "0x00",
    })
    expect(res.reverted).toBe(true)
  })
})
